### ISSUE TITLE
Cannot run programs from PATH when full path has spaces and quoted arguments are passed

### ISSUE THREAD

**Issue by dmitri-gb** [malxau/yori#24]
For example, say there is a file `C:\Program Files\Foo\bin\foo.cmd` (doesn't matter what's in it). My PATH contains the directory `C:\Program Files\Foo\bin`. Now, if I just invoke `foo`, then it works as expected. Similarly, if I invoke, say, `foo --help` then that works as well. However, if I try to invoke `foo "--help"` then it fails with

```
'C:\Program' is not recognized as an internal or external command, operable program or batch file.
```

It looks like adding quoted arguments makes Yori unable to run programs (or at least CMD scripts) that have spaces in their full path. This happens with any path that has spaces, not just `C:\Program Files`.

Of course, all works as expected with vanilla cmd.exe.

**Comment by malxau** [comment]
I'm tempted to say that this is a bug in CMD, although I realize that's unhelpful.  I'll try messing around with this when I have some spare time to see what it's willing to accept.

This bug occurs because foo.cmd is a CMD script, so Yori is trying to construct a command to execute CMD with the specified script.  In this case, it will build:
```
C:\Windows\system32\cmd.exe /c "C:\Program Files\Foo\bin\foo.cmd" "--help"
```

If you type this command into CMD, you'll see the error above.  The error is coming from CMD, not Yori.  It looks like CMD is taking everything following the "/c" as a single string, and parsing it again, losing the benefit of the quotes.  So it tries to execute:
```
C:\Program Files\Foo\bin\foo.cmd --help
```
which confuses CMD because it can't resolve the script name.

What's odd is that
```
C:\Windows\system32\cmd.exe /c "C:\Program Files\Foo\bin\foo.cmd" --help
```
does resolve the script name correctly.  It really looks like CMD is looking for the final quote -- being greedy -- then taking the string in the middle and re-parsing it.  But if that's true, the question is how to tell it what the real script name is.

**Comment by dmitri-gb** [comment]
Aha, interesting!

The examples on [ss64.com](https://ss64.com/nt/cmd.html) suggest that in cases like this, one should add extra quotes around the whole command string. So to make my example work, Yori should execute
```
C:\Windows\system32\cmd.exe /c ""C:\Program Files\Foo\bin\foo.cmd" "--help""
```

The relevant part from CMD help is this:
```
If /C or /K is specified, then the remainder of the command line after
the switch is processed as a command line, where the following logic is
used to process quote (") characters:

    1.  If all of the following conditions are met, then quote characters
        on the command line are preserved:

        - no /S switch
        - exactly two quote characters
        - no special characters between the two quote characters,
          where special is one of: &<>()@^|
        - there are one or more whitespace characters between the
          two quote characters
        - the string between the two quote characters is the name
          of an executable file.

    2.  Otherwise, old behavior is to see if the first character is
        a quote character and if so, strip the leading character and
        remove the last quote character on the command line, preserving
        any text after the last quote character.
```
Basically, case 1 is the reason this works:
```
C:\Windows\system32\cmd.exe /c "C:\Program Files\Foo\bin\foo.cmd" --help
```
however case 1 only applies in limited situations ("exactly two quote characters"), so the aim should be to always trigger case 2. This can also be forced with the `/S` switch.

**Comment by malxau** [comment]
I just released 1.21 which I believe fixes this issue.  You can upgrade an existing install with `ypm -u`.  If you find any additional problems, please either reopen this issue or file a new one describing that problem.  Thanks for the report!
